### PR TITLE
Feature/SRV-271

### PR DESCRIPTION
**Pull request by nest1ing**

### CODE REVIEW

File: agent/data/table.hpp [hunk 1]
```diff
@@ -90,7 +90,7 @@ template <typename ItemType> class Table
         // Drop sensors if it not present in answer
         for (auto it = _items.begin(); it != _items.end();)
         {
-            if (data.find((*it)->path) == data.end())
+            if (data.find(_path + "/" + (*it)->name) == data.end())
```
Comment (artemsen) [inline agent/data/table.hpp]: Construct path (_path + "/") part outside the cycle.

File: agent/snmp_oid.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    std::vector<char> oid_str;
+
+    // Calculate required size of oid_strfer
+    va_start(ap, fmt);
+    int n = vsnprintf(nullptr, 0, fmt, ap);
+    va_end(ap);
+
+    if (n > 0)
+    {
+        oid_str.resize(n + 1); // for temrminating '\0'
```
Comment (artemsen) [inline agent/snmp_oid.hpp]: terminating?

File: agent/yadro/inventory.cpp [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            _presentOid, ".1.3.6.1.4.1.49769.4.1.%lu.\"%s\"",
+            COLUMN_YADROINVENTORY_PRESENT, name.c_str());
+        phosphor::snmp::agent::make_oid(
+            _functionalOid, ".1.3.6.1.4.1.49769.4.1.%lu.\"%s\"",
+            COLUMN_YADROINVENTORY_FUNCTIONAL, name.c_str());
+    }
+
+    void setFields(const fields_map_t& fields) override
+    {
+        bool isPresent = std::get<FIELD_INVENTORY_PRESENT>(data),
```
Comment (artemsen) [inline agent/yadro/inventory.cpp]: Bad style

File: agent/yadro/sensors.cpp [hunk 4]
```diff
@@ -140,30 +130,32 @@ struct Sensor
...
-                snprintf(stateOidStr, sizeof(stateOidStr) - 1,
-                         ".1.3.6.1.4.1.49769.1.%lu.1.7.\"%s\"",
-                         _notifyOid.back(), _name.c_str());
-                read_objid(stateOidStr, _stateOid.data(), &stateOidLen);
-                _stateOid.resize(stateOidLen);
+                case ST_TEMPERATURE:
+                case ST_VOLTAGE:
+                case ST_CURRENT:
+                case ST_POWER:
+                    _power = -3;
```
Comment (artemsen) [inline agent/yadro/sensors.cpp]: Magic number

**Comment by nest1ing** [comment]
I fixed all comments from @artemsen

File: agent/data/table.hpp (hunk 1 above)
Comment (nest1ing) [reply agent/data/table.hpp]: done

File: agent/snmp_oid.hpp (hunk 2 above)
Comment (nest1ing) [reply agent/snmp_oid.hpp]: Fixed

File: agent/Makefile.am [hunk 5]
```diff
@@ -4,10 +4,10 @@ sbin_PROGRAMS = phosphor-snmp-agent
 
 phosphor_snmp_agent_SOURCES = 	\
 		snmp.cpp 				\
-		snmptrap.cpp 			\
 		yadro/powerstate.cpp 	\
 		yadro/sensors.cpp 		\
 		yadro/software.cpp 		\
+		yadro/inventory.cpp 	\
```
Comment (artemsen) [inline agent/Makefile.am]: Test message

File: agent/data/table.hpp [hunk 6]
```diff
@@ -88,9 +88,10 @@ template <typename ItemType> class Table
         auto data = sdbusplus::helper::helper::getSubTree(_path, _interfaces);
 
         // Drop sensors if it not present in answer
+        auto path = _path + "/";
         for (auto it = _items.begin(); it != _items.end();)
```
Comment (AlexanderAmelkin) [inline agent/data/table.hpp:92]: why not `for (auto it : _items)` ?

File: agent/data/table.hpp [hunk 7]
```diff
@@ -222,8 +234,9 @@ template <typename ItemType> class Table
      */
     void dropItem(const std::string& path)
     {
-        auto it = std::lower_bound(_items.begin(), _items.end(), path);
-        if (it != _items.end() && (*it)->path == path)
+        auto name = path.substr(_path.length() + 1);
```
Comment (AlexanderAmelkin) [inline agent/data/table.hpp]: What's this `+ 1` is for? What are you doing here? A comment would help.

File: agent/data/table/item.hpp [hunk 8]
```diff
@@ -58,13 +58,15 @@ template <typename... T> struct Item
...
+     * @param name - DBus object path relative by folder
      * @param args - Default values for each fields
      */
-    Item(const std::string& p, T&&... args) :
-        path(p), data(std::forward<T>(args)...),
+    Item(const std::string& folder, const std::string& name, T&&... args) :
+        name(name), data(std::forward<T>(args)...),
         changedMatch(sdbusplus::helper::helper::getBus(),
-                     sdbusplus::bus::match::rules::propertiesChanged(path),
+                     sdbusplus::bus::match::rules::propertiesChanged(
```
Comment (AlexanderAmelkin) [inline agent/data/table/item.hpp:68]: maybe add a `using` somewhere above and then use just `propertiesChanged` here and not wrap the arguments?

File: agent/main.cpp [hunk 9]
```diff
@@ -52,9 +53,9 @@ void print_usage()
 }
 
 // parse_args error codes
-#define EC_SHOW_USAGE    1
-#define EC_ERROR        -1
-#define EC_SUCCESS       0
+#define EC_SHOW_USAGE 1
+#define EC_ERROR -1
+#define EC_SUCCESS 0
```
Comment (AlexanderAmelkin) [inline agent/main.cpp]: I'd make this an enum-type and changed `parse_args` from `int` to that type.

**Review by artemsen — DISMISSED** [review]

File: agent/snmpvars.hpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * @brief Fill snmp field with boolean value.
+     */
+    static void set(netsnmp_variable_list* var, bool value)
+    {
+        snmp_set_var_typed_integer(
+            var, ASN_INTEGER,
+            static_cast<int>(value ? TruthValue::True : TruthValue::False));
```
Comment (AlexanderAmelkin) [inline agent/snmpvars.hpp]: Please replace this cast with something like an inline function or a macro (e.g. `SNMPBOOL(value)` or `toSNMPbool(value)`.

File: agent/snmpvars.hpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+     * @param vars - Pointer to snmp variables list
+     * @param field_oid - field OID
+     * @param field_oid_len - length of field OID
+     * @param field_value - field value
+     */
+    static void add(netsnmp_variable_list* vars, const oid* field_oid,
+                    size_t field_oid_len, bool field_value)
+    {
+        auto v = static_cast<u_char>(field_value ? TruthValue::True
+                                                 : TruthValue::False);
```
Comment (AlexanderAmelkin) [inline agent/snmpvars.hpp]: same macro/inline as before.

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: agent/data/table.hpp (hunk 6 above)
Comment (nest1ing) [reply agent/data/table.hpp:92]: because `_items` maybe changed inside the loop

File: agent/data/table/item.hpp (hunk 8 above)
Comment (nest1ing) [reply agent/data/table/item.hpp:68]: for one time using? I don't think that needed.

File: agent/main.cpp (hunk 9 above)
Comment (nest1ing) [reply agent/main.cpp]: Value, returned from `parse_args`, next used as `application exit code` and it should be `int`.
In next commit I will replace these macros with  `anonymous enum`.

File: agent/snmpvars.hpp (hunk 10 above)
Comment (nest1ing) [reply agent/snmpvars.hpp]: done

**Comment by nest1ing** [comment]
I push some changes wanted by @AlexanderAmelkin.
However I would like to squash it after approve.

File: agent/snmpvars.hpp [hunk 12]
```diff
@@ -36,6 +36,9 @@ enum class TruthValue : int
     False = 2,
 };
 
+#define SNMPBOOL(value)                                                        \
```
Comment (AlexanderAmelkin) [inline agent/snmpvars.hpp]: why so many spaces before `\` ? Please put it right after the closing paren.

**Comment by AlexanderAmelkin** [comment]
I can squash everything myself when merging if you only want one commit.
Or you can squash/group the commits yourself as you see fit.

File: agent/data/table/item.hpp (hunk 8 above)
Comment (AlexanderAmelkin) [reply agent/data/table/item.hpp:68]: IMHO, that would just make it easier to read. I don't insist.

File: agent/main.cpp (hunk 9 above)
Comment (AlexanderAmelkin) [reply agent/main.cpp]: `enum`s are implicitly converted to `int` without problems. However changing the function return type to an enum would protect the function from occasionally returning an int code that is not a part of the enum. You'd see an 'invalid conversion' compiler error if you attempted to do that.

Example:

```

typedef enum {
	SOMEVAL,
	OTHERVAL
} mynum_t;


mynum_t rc(int x)
{
	if (x) return SOMEVAL;
	else return 5; // This gives 'invalid conversion' error
}

int main()
{
	return rc(2); // This works just fine and silently, even with -Wall
}
```

File: agent/snmpvars.hpp (hunk 12 above)
Comment (nest1ing) [reply agent/snmpvars.hpp]: It made by `clang-format` and it put white spaces again when i remove they.

**Comment by nest1ing** [comment]
I group commits.

**Review by AlexanderAmelkin — APPROVED** [review]

**Pull request merged**